### PR TITLE
Adopt dynamicDowncast<> in editing's HTMLConverter

### PR DESCRIPTION
**Pull request by annevk**
#### 65b42b723b10db32162e2567cd7a54136149a69b
<pre>
Adopt dynamicDowncast&lt;&gt; in editing&apos;s HTMLConverter
<a href="https://bugs.webkit.org/show_bug.cgi?id=268442">https://bugs.webkit.org/show_bug.cgi?id=268442</a>

Reviewed by Chris Dumez.

* Source/WebCore/editing/cocoa/HTMLConverter.mm:
(HTMLConverterCaches::inlineStylePropertyForElement):
(stringFromCSSValue):
(HTMLConverterCaches::propertyValueForNode):
(HTMLConverterCaches::floatPropertyValueForNode):
(HTMLConverter::_blockLevelElementForNode):
(HTMLConverterCaches::colorPropertyValueForNode):
(HTMLConverter::_addTableCellForElement):
(HTMLConverter::_processElement):
(HTMLConverter::_traverseNode):
(fileWrapperForElement):
(WebCore::editingAttributedString):

Canonical link: <a href="https://commits.webkit.org/273881@main">https://commits.webkit.org/273881@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/381ce987ec2d73abea633e2593a6d302ebcb6734

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/36870 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/15803 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/39137 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/39507 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/33013 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/38155 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/18320 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/12920 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/31567 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/37432 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/13347 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/32577 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/11652 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/11675 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/32869 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/40757 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/33447 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/33211 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/37572 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/11951 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/9760 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/35701 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/13609 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8374 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/12341 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/12859 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/editing/cocoa/HTMLConverter.mm [hunk 1]
```diff
@@ -480,72 +485,73 @@ static bool stringFromCSSValue(CSSValue& value, String& result)
 
     switch (propertyId) {
     case CSSPropertyDisplay:
-        if (element.hasTagName(headTag) || element.hasTagName(scriptTag) || element.hasTagName(appletTag) || element.hasTagName(noframesTag))
+        if (element->hasTagName(headTag) || element->hasTagName(scriptTag) || element->hasTagName(appletTag) || element->hasTagName(noframesTag))
```
Comment (annevk) [inline Source/WebCore/editing/cocoa/HTMLConverter.mm]: Should we make this a `switch(element->nodeName())`? Wasn't sure how much additional refactoring I should throw in. Could also do it as a follow-up.

File: Source/WebCore/editing/cocoa/HTMLConverter.mm [hunk 2]
```diff
@@ -456,21 +461,21 @@ static bool stringFromCSSValue(CSSValue& value, String& result)
...
+    if (!element) {
+        if (auto* parent = node.parentInComposedTree())
             return propertyValueForNode(*parent, propertyId);
         return String();
     }
 
     bool inherit = false;
-    Element& element = downcast<Element>(node);
-    if (RefPtr<CSSValue> value = computedStylePropertyForElement(element, propertyId)) {
+    if (RefPtr<CSSValue> value = computedStylePropertyForElement(*element, propertyId)) {
```
Comment (cdumez) [inline Source/WebCore/editing/cocoa/HTMLConverter.mm]: Can drop `<CSSValue>`

File: Source/WebCore/editing/cocoa/HTMLConverter.mm [hunk 3]
```diff
@@ -456,21 +461,21 @@ static bool stringFromCSSValue(CSSValue& value, String& result)
...
         String result;
         if (stringFromCSSValue(*value, result))
             return result;
     }
 
-    if (RefPtr<CSSValue> value = inlineStylePropertyForElement(element, propertyId)) {
+    if (RefPtr<CSSValue> value = inlineStylePropertyForElement(*element, propertyId)) {
```
Comment (cdumez) [inline Source/WebCore/editing/cocoa/HTMLConverter.mm]: Can drop `<CSSValue>`

File: Source/WebCore/editing/cocoa/HTMLConverter.mm (hunk 1 above)
Comment (cdumez) [reply Source/WebCore/editing/cocoa/HTMLConverter.mm]: Given the number of checks here, I think switching to a `switch(element->nodeName())` is a no-brainer. Feel free to do in this patch or as a follow-up.

File: Source/WebCore/editing/cocoa/HTMLConverter.mm [hunk 4]
```diff
@@ -604,21 +610,21 @@ static inline bool floatValueFromPrimitiveValue(CSSPrimitiveValue& primitiveValu
 
 bool HTMLConverterCaches::floatPropertyValueForNode(Node& node, CSSPropertyID propertyId, float& result)
 {
-    if (!is<Element>(node)) {
-        if (ContainerNode* parent = node.parentInComposedTree())
+    RefPtr element = dynamicDowncast<Element>(node);
+    if (!element) {
+        if (auto* parent = node.parentInComposedTree())
```
Comment (cdumez) [inline Source/WebCore/editing/cocoa/HTMLConverter.mm]: We likely want to ref the parent here.

File: Source/WebCore/editing/cocoa/HTMLConverter.mm [hunk 5]
```diff
@@ -638,7 +644,7 @@ static inline bool floatValueFromPrimitiveValue(CSSPrimitiveValue& primitiveValu
     }
 
     if (inherit) {
-        if (ContainerNode* parent = node.parentInComposedTree())
+        if (auto* parent = node.parentInComposedTree())
```
Comment (cdumez) [inline Source/WebCore/editing/cocoa/HTMLConverter.mm]: We likely want to ref the parent here.

File: Source/WebCore/editing/cocoa/HTMLConverter.mm [hunk 6]
```diff
@@ -721,7 +727,9 @@ static inline bool floatValueFromPrimitiveValue(CSSPrimitiveValue& primitiveValu
 
 Element* HTMLConverter::_blockLevelElementForNode(Node* node)
 {
-    Element* element = is<Element>(node) ? downcast<Element>(node) : node->parentElement();
+    auto* element = dynamicDowncast<Element>(node);
```
Comment (cdumez) [inline Source/WebCore/editing/cocoa/HTMLConverter.mm]: We should probably ref the element here.

File: Source/WebCore/editing/cocoa/HTMLConverter.mm [hunk 7]
```diff
@@ -744,22 +752,22 @@ static Color normalizedColor(Color color, bool ignoreDefaultColor, Element& elem
 
 Color HTMLConverterCaches::colorPropertyValueForNode(Node& node, CSSPropertyID propertyId)
 {
-    if (!is<Element>(node)) {
-        if (Node* parent = node.parentInComposedTree())
+    RefPtr element = dynamicDowncast<Element>(node);
+    if (!element) {
+        if (auto* parent = node.parentInComposedTree())
```
Comment (cdumez) [inline Source/WebCore/editing/cocoa/HTMLConverter.mm]: RefPtr

File: Source/WebCore/editing/cocoa/HTMLConverter.mm [hunk 8]
```diff
@@ -781,7 +789,7 @@ static Color normalizedColor(Color color, bool ignoreDefaultColor, Element& elem
     }
 
     if (inherit) {
-        if (Node* parent = node.parentInComposedTree())
+        if (auto* parent = node.parentInComposedTree())
```
Comment (cdumez) [inline Source/WebCore/editing/cocoa/HTMLConverter.mm]: RefPtr

File: Source/WebCore/editing/cocoa/HTMLConverter.mm [hunk 9]
```diff
@@ -1677,10 +1683,9 @@ static NSInteger _colCompare(id block1, id block2, void *)
             _addTableForElement(nil);
         _addTableCellForElement(&element);
 #if ENABLE(ATTACHMENT_ELEMENT)
-    } else if (is<HTMLAttachmentElement>(element)) {
-        HTMLAttachmentElement& attachment = downcast<HTMLAttachmentElement>(element);
-        if (attachment.file()) {
-            NSURL *url = [NSURL fileURLWithPath:attachment.file()->path()];
+    } else if (RefPtr attachment = dynamicDowncast<HTMLAttachmentElement>(element)) {
+        if (attachment->file()) {
+            NSURL *url = [NSURL fileURLWithPath:attachment->file()->path()];
```
Comment (cdumez) [inline Source/WebCore/editing/cocoa/HTMLConverter.mm]: Would be good to use RetainPtr here.

File: Source/WebCore/editing/cocoa/HTMLConverter.mm [hunk 10]
```diff
@@ -1722,8 +1727,8 @@ static NSInteger _colCompare(id block1, id block2, void *)
             if (url)
                 retval = !_addAttachmentForElement(element, url, isBlockLevel, NO);
         }
-    } else if (is<HTMLFrameElementBase>(element)) {
-        if (Document* contentDocument = downcast<HTMLFrameElementBase>(element).contentDocument()) {
+    } else if (auto* frameElement = dynamicDowncast<HTMLFrameElementBase>(element)) {
+        if (auto* contentDocument = frameElement->contentDocument()) {
```
Comment (cdumez) [inline Source/WebCore/editing/cocoa/HTMLConverter.mm]: RefPtr

File: Source/WebCore/editing/cocoa/HTMLConverter.mm [hunk 11]
```diff
@@ -1754,26 +1759,24 @@ static NSInteger _colCompare(id block1, id block2, void *)
...
     } else if (element.hasTagName(qTag)) {
         _addQuoteForElement(element, YES, _quoteLevel++);
     } else if (element.hasTagName(inputTag)) {
-        if (is<HTMLInputElement>(element)) {
-            HTMLInputElement& inputElement = downcast<HTMLInputElement>(element);
-            if (inputElement.type() == textAtom()) {
-                NSString *value = inputElement.value();
+        if (RefPtr inputElement = dynamicDowncast<HTMLInputElement>(element)) {
+            if (inputElement->type() == textAtom()) {
+                NSString *value = inputElement->value();
```
Comment (cdumez) [inline Source/WebCore/editing/cocoa/HTMLConverter.mm]: RetainPtr

File: Source/WebCore/editing/cocoa/HTMLConverter.mm [hunk 12]
```diff
@@ -1754,26 +1759,24 @@ static NSInteger _colCompare(id block1, id block2, void *)
...
                 if (value && [value length] > 0)
                     _addValue(value, element);
             }
         }
     } else if (element.hasTagName(textareaTag)) {
-        if (is<HTMLTextAreaElement>(element)) {
-            HTMLTextAreaElement& textAreaElement = downcast<HTMLTextAreaElement>(element);
-            NSString *value = textAreaElement.value();
+        if (RefPtr textAreaElement = dynamicDowncast<HTMLTextAreaElement>(element)) {
+            NSString *value = textAreaElement->value();
```
Comment (cdumez) [inline Source/WebCore/editing/cocoa/HTMLConverter.mm]: RetainPtr

File: Source/WebCore/editing/cocoa/HTMLConverter.mm [hunk 13]
```diff
@@ -2232,8 +2234,8 @@ static NSInteger _colCompare(id block1, id block2, void *)
     }
 
     auto* renderer = element.renderer();
-    if (is<RenderImage>(renderer)) {
-        auto* image = downcast<RenderImage>(*renderer).cachedImage();
+    if (auto* renderImage = dynamicDowncast<RenderImage>(renderer)) {
+        auto* image = renderImage->cachedImage();
```
Comment (cdumez) [inline Source/WebCore/editing/cocoa/HTMLConverter.mm]: `CachedResourceHandle`

**Review by cdumez — APPROVED** [review]
LGTM with nits.

File: Source/WebCore/editing/cocoa/HTMLConverter.mm [hunk 14]
```diff
@@ -1677,12 +1730,11 @@ static NSInteger _colCompare(id block1, id block2, void *)
             _addTableForElement(nil);
         _addTableCellForElement(&element);
 #if ENABLE(ATTACHMENT_ELEMENT)
-    } else if (is<HTMLAttachmentElement>(element)) {
-        HTMLAttachmentElement& attachment = downcast<HTMLAttachmentElement>(element);
-        if (attachment.file()) {
-            NSURL *url = [NSURL fileURLWithPath:attachment.file()->path()];
+    } else if (RefPtr attachment = dynamicDowncast<HTMLAttachmentElement>(element)) {
+        if (attachment->file()) {
+            RetainPtr<NSURL> url = [NSURL fileURLWithPath:attachment->file()->path()];
```
Comment (cdumez) [inline Source/WebCore/editing/cocoa/HTMLConverter.mm]: Probably don't need `<NSURL>`?

**Review by cdumez — APPROVED** [review]
LGTM

**Pull request merged**